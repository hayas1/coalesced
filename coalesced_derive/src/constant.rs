use std::cell::LazyCell;

use syn::{parse_quote, Ident, Path};

pub const DERIVE_CONSTRUCTION: &str = "Construction";

thread_local! {
    pub static PATH_SEMIGROUP: LazyCell<Path> = LazyCell::new(|| parse_quote!(::coalesced::Semigroup));
    pub static IDENT_SEMIGROUP_OP: LazyCell<Ident> = LazyCell::new(|| parse_quote!(semigroup_op));
    pub static PATH_ANNOTATED: LazyCell<Path> = LazyCell::new(|| parse_quote!(::coalesced::Annotated));
    pub static PATH_REVERSED: LazyCell<Path> =
        LazyCell::new(|| parse_quote!(::coalesced::op::reverse::Reversed));
}
