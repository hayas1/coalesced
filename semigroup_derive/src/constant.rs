use syn::{parse_quote, Attribute, Path, TypeParam};

pub const DERIVE_CONSTRUCTION: &str = "Construction";
pub const DERIVE_SEMIGROUP: &str = "Semigroup";

#[derive(Debug, Clone)]
pub struct Constant {
    pub path_semigroup: Path,
    pub path_annotated_semigroup: Path,
    pub path_annotated: Path,
    pub path_annotate: Path,
    pub path_monoid: Path,
    pub path_commutative: Path,
    pub path_construction_trait: Path,
    pub path_construction_annotated: Path,
    pub path_construction_monoid: Path,
    pub default_type_param: TypeParam,
    pub attr_feature_monoid: Option<Attribute>,
    pub attr_feature_async: Option<Attribute>,
}
pub trait ConstantExt {
    fn constant() -> Constant;
}
pub enum External {}
impl ConstantExt for External {
    fn constant() -> Constant {
        Constant {
            path_semigroup: parse_quote! {::semigroup::Semigroup},
            path_annotated_semigroup: parse_quote! {::semigroup::AnnotatedSemigroup},
            path_annotated: parse_quote! {::semigroup::Annotated},
            path_annotate: parse_quote! {::semigroup::Annotate},
            path_monoid: parse_quote! {::semigroup::Monoid},
            path_commutative: parse_quote! {::semigroup::Commutative},
            path_construction_trait: parse_quote! {::semigroup::Construction},
            path_construction_annotated: parse_quote! {::semigroup::ConstructionAnnotated},
            path_construction_monoid: parse_quote! {::semigroup::ConstructionMonoid},
            default_type_param: parse_quote! { A },
            attr_feature_monoid: None,
            attr_feature_async: None,
        }
    }
}
pub enum Internal {}
impl ConstantExt for Internal {
    fn constant() -> Constant {
        Constant {
            path_semigroup: parse_quote! {crate::Semigroup},
            path_annotated_semigroup: parse_quote! {crate::AnnotatedSemigroup},
            path_annotated: parse_quote! {crate::Annotated},
            path_annotate: parse_quote! {crate::Annotate},
            path_monoid: parse_quote! {crate::Monoid},
            path_commutative: parse_quote! {crate::Commutative},
            path_construction_trait: parse_quote! {crate::Construction},
            path_construction_annotated: parse_quote! {crate::ConstructionAnnotated},
            path_construction_monoid: parse_quote! {crate::ConstructionMonoid},
            default_type_param: parse_quote! { A },
            attr_feature_monoid: Some(parse_quote! {#[cfg(feature = "monoid")]}),
            attr_feature_async: Some(parse_quote! {#[cfg(feature = "async")]}),
        }
    }
}
