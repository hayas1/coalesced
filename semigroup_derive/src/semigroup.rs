use proc_macro2::TokenStream;
use quote::ToTokens;
use syn::DeriveInput;

use crate::{
    constant::ConstantExt,
    semigroup::{ast::Semigroup, attr::ContainerAttr},
};

mod ast;
mod attr;

pub fn impl_semigroup<C: ConstantExt>(derive: &DeriveInput) -> syn::Result<TokenStream> {
    let constant = C::constant();
    let attr = ContainerAttr::new(derive)?;
    let semigroup = Semigroup::new(&constant, derive, &attr)?;
    Ok(semigroup.into_token_stream())
}

#[cfg(test)]
mod tests {
    use rstest::rstest;

    use crate::constant::{External, Internal};

    use super::*;

    #[rstest]
    #[case::semigroup_annotated(
        "semigroup_annotated",
        impl_semigroup::<External>,
        syn::parse_quote! {
            #[derive(Semigroup)]
            #[semigroup(annotated)]
            pub struct NamedStruct {
                #[semigroup(with = "semigroup::op::Overwrite")]
                pub foo: String,
                pub bar: Option<u32>,
                pub baz: semigroup::op::Overwrite<bool>,
            }
        },
    )]
    #[case::semigroup_not_annotated(
        "semigroup_not_annotated",
        impl_semigroup::<Internal>,
        syn::parse_quote! {
            #[derive(SemigroupPriv)]
            #[semigroup(with = "semigroup::op::Overwrite")]
            pub struct UnnamedStruct<T: std::ops::Add> (
                #[semigroup(with = "semigroup::op::Added")]
                T,
                u64
            );
        },
    )]
    #[case::semigroup_custom_annotation(
        "semigroup_custom_annotation",
        impl_semigroup::<External>,
        syn::parse_quote! {
            #[derive(Semigroup)]
            #[semigroup(annotated, annotation_param = X, with = "semigroup::op::Overwrite")]
            pub struct NamedStruct{
                pub foo: String,
                pub bar: Option<u32>,
                pub baz: bool,
            }
        },
    )]
    #[case::semigroup_monoid(
        "semigroup_monoid",
        impl_semigroup::<External>,
        syn::parse_quote! {
            #[derive(Semigroup)]
            #[semigroup(monoid, commutative)]
            pub struct MonoidStruct{
                #[semigroup(with = "semigroup::op::Sum")]
                pub sum: u32,
            }
        },
    )]
    fn test_derive_semigroup_snapshot(
        #[case] case: &str,
        #[case] f: impl Fn(&DeriveInput) -> syn::Result<TokenStream>,
        #[case] input: DeriveInput,
    ) {
        let generated = f(&input).unwrap();
        let formatted = prettyplease::unparse(&syn::parse2(generated).unwrap());
        insta::with_settings!({ snapshot_path => "../tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_snapshot!(case, formatted);
        });
    }
}
