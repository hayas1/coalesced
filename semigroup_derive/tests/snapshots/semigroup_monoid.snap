---
source: semigroup_derive/src/semigroup.rs
assertion_line: 89
expression: formatted
---
#[automatically_derived]
impl ::semigroup::Semigroup for MonoidStruct {
    fn op(base: Self, other: Self) -> Self {
        Self {
            sum: <semigroup::op::semigroup::sum::Sum<
                _,
            > as ::semigroup::Construction<_>>::lift_op(base.sum, other.sum),
        }
    }
}
#[automatically_derived]
impl ::semigroup::Monoid for MonoidStruct
where
    Self: Default,
{
    fn unit() -> Self {
        Default::default()
    }
}
#[automatically_derived]
impl ::semigroup::Commutative for MonoidStruct {}
